src/Libraries/Microsoft.Extensions.AI.Evaluation.NLP/SimpleWordTokenizer.cs

```diff
@@ -0,0 +1,208 @@
+    /// if the input text contains no valid tokens.</returns>
+    public static IEnumerable<string> WordTokenize(ReadOnlyMemory<char> text)
+    {
+        StringBuilder sb = new StringBuilder();
+
+        while (true)
+        {
+            if (text.IsEmpty)
+            {
+                if (sb.Length > 0)
+                {
+                    yield return sb.ToString();
+                    _ = sb.Clear();
+                }
+
+                yield break;
+            }
+
+            var span = text.Span;
+            char nextChar = span[0];
+
+            // Skip whitespace as separator
+            if (char.IsWhiteSpace(nextChar))
+            {
+                if (sb.Length > 0)
+                {
+                    yield return sb.ToString();
+                    _ = sb.Clear();
+                }
+
+                text = text.Slice(1);
+                continue;
+            }
+
+            // Join hyphenated words
+            if (span[0] == '-' &&
+                span.Length > 1 &&
+                span[1] == '\n')
+            {
+#pragma warning disable S109 // Magic numbers should not be used
+                text = text.Slice(2);
+#pragma warning restore S109 // Magic numbers should not be used
+                continue;
+            }
+
+            // Translate HTML entities
+            if (nextChar == '&')
+            {
+                if (span.StartsWith("&quot;".AsSpan()))
+                {
+                    if (sb.Length > 0)
+                    {
+                        yield return sb.ToString();
+                        _ = sb.Clear();
+                    }
+
+                    text = text.Slice("&quot;".Length);
+                    yield return "\"";
+                    continue;
+                }
+                else if (span.StartsWith("&amp;".AsSpan()))
+                {
+                    if (sb.Length > 0)
+                    {
+                        yield return sb.ToString();
+                        _ = sb.Clear();
+                    }
+
+                    text = text.Slice("&amp;".Length);
+                    yield return "&";
+                    continue;
+                }
+                else if (span.StartsWith("&lt;".AsSpan()))
+                {
+                    if (sb.Length > 0)
+                    {
+                        yield return sb.ToString();
+                        _ = sb.Clear();
+                    }
+
+                    text = text.Slice("&lt;".Length);
+                    yield return "<";
+                    continue;
+                }
+                else if (span.StartsWith("&gt;".AsSpan()))
+                {
+                    if (sb.Length > 0)
+                    {
+                        yield return sb.ToString();
+                        _ = sb.Clear();
+                    }
+
+                    text = text.Slice("&gt;".Length);
+                    yield return ">";
+                    continue;
+                }
```

Comment:
Is possible to extract a helper since all of these else blocks are repeating the same logic for a different character / character sequence?

I guess the `yield` might make it complicated...